libraries/rush-lib/src/logic/operations/CacheableOperationPlugin.ts

```diff
@@ -0,0 +1,522 @@
+    hooks.operationExecutionManager.tap(
+      PLUGIN_NAME,
+      (operationExecutionManager: OperationExecutionManager) => {
+        operationExecutionManager.hooks.afterExecuteOperation.tapPromise(
+          PLUGIN_NAME,
+          async (operation: OperationExecutionRecord): Promise<OperationExecutionRecord> => {
+            const { runner, status, consumers } = operation;
+            const buildCacheContext: IOperationBuildCacheContext | undefined =
+              this._getBuildCacheContextByRunner(runner);
+
+            let blockCacheWrite: boolean = !buildCacheContext?.isCacheWriteAllowed;
+            let blockSkip: boolean = !buildCacheContext?.isSkipAllowed;
+
+            switch (status) {
+              case OperationStatus.Skipped: {
+                // Skipping means cannot guarantee integrity, so prevent cache writes in dependents.
+                blockCacheWrite = true;
+                break;
+              }
+
+              case OperationStatus.SuccessWithWarning:
+              case OperationStatus.Success: {
+                // Legacy incremental build, if asked, prevent skip in dependents if the operation executed.
+                blockSkip ||= !operationExecutionManager.changedProjectsOnly;
+                break;
+              }
+            }
+
+            // Apply status changes to direct dependents
+            for (const item of consumers) {
+              const itemRunnerBuildCacheContext: IOperationBuildCacheContext | undefined =
+                this._getBuildCacheContextByRunner(item.runner);
+              if (itemRunnerBuildCacheContext) {
+                if (blockCacheWrite) {
+                  itemRunnerBuildCacheContext.isCacheWriteAllowed = false;
+                }
+                if (blockSkip) {
+                  itemRunnerBuildCacheContext.isSkipAllowed = false;
+                }
+              }
+            }
+            return operation;
+          }
+        );
+      }
+    );
+
+    hooks.afterExecuteOperations.tapPromise(PLUGIN_NAME, async () => {
+      this._buildCacheContextByOperationRunner.clear();
+    });
+  }
+
+  private _applyOperationRunner(runner: ShellOperationRunner, context: ICreateOperationsContext): void {
+    const { buildCacheConfiguration, cobuildConfiguration } = context;
+    const { hooks } = runner;
+
+    const buildCacheContext: IOperationBuildCacheContext = this._getBuildCacheContextByRunnerOrThrow(runner);
+
+    hooks.beforeExecute.tapPromise(
+      PLUGIN_NAME,
+      async (beforeExecuteContext: IOperationRunnerBeforeExecuteContext) => {
+        const earlyReturnStatus: OperationStatus | undefined = await (async () => {
+          const {
+            context,
+            runner,
+            terminal,
+            lastProjectDeps,
+            projectDeps,
+            trackedProjectFiles,
+            logPath,
+            errorLogPath,
+            rushProject,
+            phase,
+            selectedPhases,
+            projectChangeAnalyzer,
```

Comment:
These two values (`selectedPhases`, `projectChangeAnalyzer`) are constant across all operations and are part of the `ICreateOperationsContext` object passed to the `createOperations` hook, so rather than plumbing them through the runner you could stash them in your `IOperationBuildCacheContext`, since this is the only plugin that uses them.